### PR TITLE
DT-510 queue health

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/health/IntegrationTest.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  @Qualifier("awsSqsClient")
+  internal lateinit var awsSqsClient: AmazonSQS
+
+  companion object {
+    internal val eliteMockServer = EliteMockServer()
+
+    @BeforeAll
+    @JvmStatic
+    fun startMocks() {
+      eliteMockServer.start()
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/health/IntegrationTest.kt]: junit 5 doesn't have class rules, so have to do it manually.  They haven't got around to adding support to wiremock for junit 5 yet either

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/config/JmsConfig.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+          .withCredentials(AWSStaticCredentialsProvider(AnonymousAWSCredentials()))
+          .build()
+
+  @Bean("awsSqsDlqClient")
+  @ConditionalOnExpression("'\${sqs.provider}'.equals('localstack') and '\${sqs.embedded}'.equals('false')")
+  open fun awsSqsDlqClientLocalstack(@Value("\${sqs.endpoint.url}") serviceEndpoint: String,
+                                     @Value("\${sqs.endpoint.region}") region: String): AmazonSQS =
+      AmazonSQSClientBuilder.standard()
+          .withEndpointConfiguration(AwsClientBuilder.EndpointConfiguration(serviceEndpoint, region))
+          .withCredentials(AWSStaticCredentialsProvider(AnonymousAWSCredentials()))
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/config/JmsConfig.kt]: Nice, I assume that means you don't have to provide dummy credentials

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/QueueHealth.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    val log: Logger = LoggerFactory.getLogger(this::class.java)
+  }
+
+  override fun health(): Health {
+    val queueAttributes = try {
+      val url = awsSqsClient.getQueueUrl(queueName)
+      awsSqsClient.getQueueAttributes(getQueueAttributesRequest(url))
+    } catch (e: Exception) {
+      log.error("Unable to retrieve queue attributes for queue '{}' due to exception:", queueName, e)
+      return Builder().down().withException(e).build()
```
Comment (andymarke) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/QueueHealth.kt]: Kotlin question; I assume the `return` breaks out of parent block? And without the return the result of the Builder() would be assigned to `val queueAttributes` ?

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  @Bean
+  open fun localStackContainer(): LocalStackContainer {
+    log.info("Starting embedded localstack...")
+    val localStackContainer: LocalStackContainer = LocalStackContainer()
+        .withServices(LocalStackContainer.Service.SQS)
+        .withEnv("HOSTNAME_EXTERNAL", "localhost")
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt]: what does this do? `HOSTNAME_EXTERNAL`

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/health/HealthCheckIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+import uk.gov.justice.digital.hmpps.keyworker.services.health.QueueAttributes.*
+
+class HealthCheckIntegrationTest : IntegrationTest() {
+
+  @Autowired
+  private lateinit var queueHealth: QueueHealth
+
+  @Autowired
+  @Value("\${sqs.queue.name}")
+  @Suppress("SpringJavaInjectionPointsAutowiringInspection")
```
Comment (andymarke) [inline src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/health/HealthCheckIntegrationTest.kt]: Nice, I wondered how to get rid of those stupid warnings

Comment (petergphillips) [reply src/test/java/uk/gov/justice/digital/hmpps/keyworker/services/health/HealthCheckIntegrationTest.kt]: intellij provides this as a suggestion on the warning

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt (hunk 4 above)
Comment (petergphillips) [reply src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt]: no idea, @mikehalmamoj wrote it originally :grinning:

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/QueueHealth.kt (hunk 3 above)
Comment (petergphillips) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/health/QueueHealth.kt]: agreed

File: src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt (hunk 4 above)
Comment (mikehalmamoj) [reply src/test/java/uk/gov/justice/digital/hmpps/keyworker/config/LocalStackConfig.kt]: It's mentioned here: https://github.com/localstack/localstack#configurations.  Apparently it default to localhost anyway :man_facepalming: .  I can't remember if I put it there for a reason or just copied and pasted?

**Pull request merged**